runsc/boot/mount_hints.go

```diff
@@ -113,6 +126,65 @@ func NewPodMountHints(spec *specs.Spec) (*PodMountHints, error) {
 	return &PodMountHints{Mounts: mnts}, nil
 }
 
+// FindMount finds the MountHint that applies to this mount.
+func (p *PodMountHints) FindMount(mountSrc string) *MountHint {
+	for _, m := range p.Mounts {
+		if m.Mount.Source == mountSrc {
+			return m
+		}
+	}
+	return nil
+}
+
+// ContainerMountHints contains a collection of mountHints for the container.
+type ContainerMountHints struct {
+	Mounts map[string]*MountHint `json:"mounts"`
+}
```

Comment:
I think this part needs more thinking.

Is our goal to add EROFS support for any arbitrary mount (non-rootfs) in a container? The answer seems like a yes. We have the EROFS driver so why not.

We can't use PodMountHints because usually container rootfs is not ready until container create time, which maybe be much later than pod create time in k8s. So we can't set EROFS annotations at pod create time. OK so we need different kind of annotations like you have added.

If you see pod mount annotations, they are identified with the mount source (see `FindMount()`). This is because multiple containers can mount the same volume at various locations. All of those mounts have 1 thing in common, the source directory.

It seems like the `ContainerMountHints` you added is taking a different approach; it is identifying by mount destination. This seems OK for rootfs, but what if a container is mounting the same EROFS image at different locations? It is unusual though to mount the same filesystem tree at various locations in the same container, but the OCI container spec allows for this. We want to *share* the same filesystem instance in that case (so that the dentry tree is shared and memory usage is lower).

Another thing to note is that when EROFS is being used, we should not expect the filesystem to be mounted on host (because why mount it twice, once on host and in gVisor). So what is `spec.Root.Path` set in your container spec? What will the `spec.Mounts[i].Source` be set to for mounts that want to use EROFS?

I suggest that you do the following:
1. Keep the same annotation structure as pod mount annotations (same fields, same requirements and same identification mechanism).
2. Update the container spec (`spec.Root.Path` and `spec.Mounts[i].Source`) to point to the image file. (I am curious in your use case what it was pointing to otherwise)

So your annotations will look something like:
```
"annotations" = [
    "dev.gvisor.spec.container.mount._internal_rootfs.type" = "erofs",
    "dev.gvisor.spec.container.mount._internal_rootfs.share" = "container",
    "dev.gvisor.spec.container.mount._internal_rootfs.source" = "/var/cont1/rootfs.img",
],
"root": {
    "path": "/var/cont1/rootfs.img",
  },
```

NOTE: I am using `_internal_rootfs` above so that the volume name does not collide with a legit volume name in k8s. But we don't need to worry about this now, since we are only supporting rootfs in container mount annotations.

Later, when we add support for other submounts with EROFS, you could even implement mount sharing with EROFS. What do you think?

Reply:
Actually I don't like the above proposal. It creates some issues around configuring filestore files with rootfs. I think at the core the discrepancy is that MountHint is built to configure mount of type [`specs.Mount`](https://github.com/opencontainers/runtime-spec/blob/6331715093bfcf25137411bfacb403235ed7d018/specs-go/config.go#L161). However, rootfs is not defined using that. The spec defines Rootfs using [just a path](https://github.com/opencontainers/runtime-spec/blob/6331715093bfcf25137411bfacb403235ed7d018/specs-go/config.go#L155) (no type).

We configure filestore file for submounts based on `spec.Mount[i].Type` and `MountHint.Mount.Type`. We can't do that for rootfs. I think we should have separate annotation for rootfs.
```
"annotations" = [
    "dev.gvisor.spec.rootfs.type" = "erofs/bind",
    "dev.gvisor.spec.rootfs.image" = "/var/cont1/rootfs.img",
    "dev.gvisor.spec.rootfs.overlay" = "none/memory/self/dir=",
],
```

Do you have any submount use case for EROFS yet?

Reply:
> Actually I don't like the above proposal. It creates some issues around configuring filestore files with rootfs. I think at the core the discrepancy is that MountHint is built to configure mount of type [`specs.Mount`](https://github.com/opencontainers/runtime-spec/blob/6331715093bfcf25137411bfacb403235ed7d018/specs-go/config.go#L161). However, rootfs is not defined using that. The spec defines Rootfs using [just a path](https://github.com/opencontainers/runtime-spec/blob/6331715093bfcf25137411bfacb403235ed7d018/specs-go/config.go#L155) (no type).
> 
> We configure filestore file for submounts based on `spec.Mount[i].Type` and `MountHint.Mount.Type`. We can't do that for rootfs. I think we should have separate annotation for rootfs.
> 
> ```
> "annotations" = [
>     "dev.gvisor.spec.rootfs.type" = "erofs/bind",
>     "dev.gvisor.spec.rootfs.image" = "/var/cont1/rootfs.img",
>     "dev.gvisor.spec.rootfs.overlay" = "none/memory/self/dir=",
> ],
> ```

I like this proposal. My first version is similar to this one (but less flexible). I also think we should have separate annotation for rootfs.

> Do you have any submount use case for EROFS yet?

In our internal use case (FaaS), user code is mounted as a separate image, but it's not mounted as a submount in `config.json`, instead, it's mounted at runtime via a [RPC](https://github.com/google/gvisor/blob/ba53672288fb20c6ded46010c148ff6fadbe4556/runsc/boot/controller.go#L705) initiated by a controller running on the same host. So I don't have any submount use case for EROFS yet.

Reply:
> In our internal use case (FaaS), user code is mounted as a separate image, but it's not mounted as a submount in config.json, instead, it's mounted at runtime via a [RPC](https://github.com/google/gvisor/blob/ba53672288fb20c6ded46010c148ff6fadbe4556/runsc/boot/controller.go#L705) initiated by a controller running on the same host. So I don't have any submount use case for EROFS yet.

Got it. When need arises, I think we should be able to reuse the existing `MountHint` as described in the first comment to add EROFS support for submounts.

nit: Actually can you update `"dev.gvisor.spec.rootfs.image"` to `"dev.gvisor.spec.rootfs.source"`? It is more consistent with the terminology used everywhere else and hence is more extendible (i.e. source can be provided with type=bind too; image is a special type for erofs only). Also I think while mounting erofs via `mount(1)` on Linux, the image path is passed as source.

Reply:
> Got it. When need arises, I think we should be able to reuse the existing `MountHint` as described in the first comment to add EROFS support for submounts.

Yeah, I agree too.

> nit: Actually can you update `"dev.gvisor.spec.rootfs.image"` to `"dev.gvisor.spec.rootfs.source"`? It is more consistent with the terminology used everywhere else and hence is more extendible (i.e. source can be provided with type=bind too; image is a special type for erofs only). Also I think while mounting erofs via `mount(1)` on Linux, the image path is passed as source.

Makes sense. Will do.